ui-trackoccupancydiagram/src/components/helpers/drawElements/drawOccupancyZones.ts

```diff
@@ -2,60 +2,290 @@ import { drawOccupancyZonesTexts } from './drawOccupancyZonesTexts';
     const trackTranslate = index === 0 ? CANVAS_PADDING : TRACK_HEIGHT_CONTAINER;
     ctx.translate(0, trackTranslate);
 
-    const trackOccupancyZones = occupancyZones?.filter((zone) => zone.trackId === track.id);
+    const filteredOccupancyZones = sortedOccupancyZones.filter((zone) => zone.trackId === track.id);
+
+    let primaryArrivalTimePixel = 0;
+    let primaryDepartureTimePixel = 0;
+    let lastDepartureTimePixel = primaryDepartureTimePixel;
+    let yPosition = OCCUPANCY_ZONE_Y_START;
+    let yOffset = 0;
+    let zoneCounter = 0;
+    let zoneIndex = 0;
+
+    while (zoneIndex < filteredOccupancyZones.length) {
+      const zone = filteredOccupancyZones[zoneIndex];
+      const arrivalTimePixel = getTimePixel(zone.arrivalTime.getTime());
+      const departureTimePixel = getTimePixel(zone.departureTime.getTime());
+      const isThroughTrain = arrivalTimePixel === departureTimePixel;
 
-    if (!trackOccupancyZones) return;
+      // * if the zone is not overlapping with any previous one, draw it in the center of the track
+      // * and reset the primary values
+      // *
+      // * if the zone is overlapping with the previous one, draw it below or above the previous one
+      // * depending on the overlapping counter
+      // *
+      // * if the zone is overlapping with the previous one and the counter is higher than the max zones
+      // * draw the remaining trains box
+      // *
+      if (arrivalTimePixel > lastDepartureTimePixel) {
+        // reset to initial value if the zone is not overlapping
+        yPosition = OCCUPANCY_ZONE_Y_START;
+        primaryArrivalTimePixel = arrivalTimePixel;
+        primaryDepartureTimePixel = departureTimePixel;
+        lastDepartureTimePixel = departureTimePixel;
+        yOffset = Y_OFFSET_INCREMENT;
+        zoneCounter = 1;
 
-    trackOccupancyZones.forEach((zone) => {
-      const arrivalTime = getTimePixel(zone.arrivalTime.getTime());
-      const departureTime = getTimePixel(zone.departureTime.getTime());
+        drawOccupationZone({
+          ctx,
+          zone,
+          arrivalTimePixel,
+          departureTimePixel,
+          yPosition,
+          isThroughTrain,
+          selectedTrainId,
+        });
 
-      ctx.fillStyle = zone.color;
-      ctx.strokeStyle = COLORS.WHITE_100;
-      ctx.lineWidth = 1;
-      ctx.beginPath();
-      ctx.rect(
-        arrivalTime,
-        OCCUPANCY_ZONE_START,
-        departureTime - arrivalTime,
-        OCCUPANCY_ZONE_HEIGHT
+        zoneIndex++;
+
+        continue;
+      } else if (zoneCounter < MAX_ZONES) {
+        // figure out if the zone is overlapping with any previous one
+        // if so and it's an even index, move it to the bottom, if it's an odd index, move it to the top
+        if (arrivalTimePixel >= primaryArrivalTimePixel) {
+          if (zoneCounter % 2 === 0) {
+            yPosition -= yOffset;
+          } else {
+            yPosition += yOffset;
+          }
+        }
```

Comment:
- You can remove the else here since you add a continue above :+1: 
- From this point, you know you are handling an overlapping train, so you can remove some old comments and checks
```suggestion
      } 
      
      if (zoneCounter < MAX_ZONES) {
        // if so and it's an even index, move it to the bottom, if it's an odd index, move it to the top
        if (zoneCounter % 2 === 0) {
          yPosition -= yOffset;
        } else {
          yPosition += yOffset;
        }
```

Reply:
sure, done